pkg/controller/metrics/metrics_with_dynamic_labels.go

```diff
@@ -0,0 +1,155 @@
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/openshift/hive/apis/hive/v1/metricsconfig"
+	"github.com/openshift/hive/pkg/constants"
+)
+
+// metricsWithDynamicLabels encompasses metrics that can have optional labels.
+type metricsWithDynamicLabels interface {
+	Observe(metav1.Object, map[string]string, float64)
+	Register()
+}
+
+type dynamicLabels struct {
+	// fixedLabels defines the label keys that are always registered on the metric.
+	fixedLabels []string
+	// optionalLabels is generated dynamically based on HiveConfig.Spec.MetricsConfig.MetricsWithDynamicLabels, and
+	// represents a map of metric label keys to ClusterDeployment metadata.label keys from which values will be gleaned
+	// when the metric is observed.
+	optionalLabels map[string]string
+}
+
+// getLabelList combines the fixed and optional labels and returns the list with just the labels that can be used to
+// define the metric
+func (d *dynamicLabels) getLabelList() []string {
+	var finalList []string
+	finalList = append(finalList, d.fixedLabels...)
+	for label := range d.optionalLabels {
+		finalList = append(finalList, label)
+	}
+	return finalList
+}
+
+// getRepeatedLabel detects and returns the first optional label it encounters, that overlaps with another label
+func (d *dynamicLabels) getRepeatedLabel() string {
+	keys := sets.Set[string]{}
+	for _, val := range d.fixedLabels {
+		keys.Insert(val)
+	}
+	for key := range d.optionalLabels {
+		if keys.Has(key) {
+			return key
+		}
+		keys.Insert(key)
+	}
+	return ""
+}
+
+// buildLabels fetches the optional label values and merges the maps of fixedLabels and optionalLabels. The resultant
+// map can be used for observing the metric.
+func (d *dynamicLabels) buildLabels(fixedLabels map[string]string, obj metav1.Object) prometheus.Labels {
+	labels := make(map[string]string, len(d.fixedLabels)+len(d.optionalLabels))
+	for _, label := range d.fixedLabels {
+		if value := fixedLabels[label]; value != "" {
+			labels[label] = value
+		} else {
+			labels[label] = constants.MetricLabelDefaultValue
+		}
+	}
+	for key, value := range d.optionalLabels {
+		labels[key] = GetLabelValue(obj, value)
+	}
+	return labels
+}
+
+type CounterVecWithDynamicLabels struct {
+	*prometheus.CounterOpts
+	metric *prometheus.CounterVec
+	*dynamicLabels
+}
+
+func NewCounterVecWithDynamicLabels(counterOpts *prometheus.CounterOpts, fixedLabels []string,
+	optionalLabels map[string]string, log log.FieldLogger) *CounterVecWithDynamicLabels {
+	counterVecMetric := &CounterVecWithDynamicLabels{
+		CounterOpts: counterOpts,
+		dynamicLabels: &dynamicLabels{
+			fixedLabels:    fixedLabels,
+			optionalLabels: optionalLabels,
+		},
+	}
+	if repeatedLabel := counterVecMetric.getRepeatedLabel(); repeatedLabel != "" {
+		log.Errorf("HiveConfig.Spec.AdditionalClusterDeploymentLabels[%q] conflicts with a fixed label for the metric %s. Please rename your label.", repeatedLabel, counterOpts.Name)
```

Comment:
We agreed to make this fatal, right?. (And if you turn this into a `panic`, you don't need the `log` arg.)